File: Compose-Bookshelf-App/app/src/test/java/com/example/bookshelfapp/GetBooksWithThumbnailsUseCaseTest.kt
package com.example.bookshelfapp

import com.example.bookshelfapp.data.BooksRepository
import com.example.bookshelfapp.domain.GetBookIdUseCase
import com.example.bookshelfapp.domain.GetBooksWithThumbnailsUseCase
import com.example.bookshelfapp.domain.SetHttpsUseCase
import com.example.bookshelfapp.fake.BooksFakeDataSource
import com.example.bookshelfapp.fake.ThumbnailFakeDataSource
import com.example.bookshelfapp.ui.BooksUiModel
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.test.runTest
import org.junit.Assert
import org.junit.Before
import org.junit.Test

class GetBooksWithThumbnailsUseCaseTest {

    private lateinit var getBooksWithThumbnailsUseCase: GetBooksWithThumbnailsUseCase

    @Before
    fun createGetBooksWithThumbnailsUseCase() {
        val getBookIdUseCase = GetBookIdUseCase(
            defaultDispatcher = Dispatchers.Default
        )

        val booksRepository = BooksRepository(
            booksDataSource = BooksFakeDataSource(),
            thumbnailDataSource = ThumbnailFakeDataSource()
        )

        val setHttpsUseCase = SetHttpsUseCase(
            defaultDispatcher = Dispatchers.Default
        )

        getBooksWithThumbnailsUseCase = GetBooksWithThumbnailsUseCase(
            booksRepository = booksRepository,
            defaultDispatcher = Dispatchers.Default,
            getBookIdUseCase = getBookIdUseCase,
            setHttpsUseCase = setHttpsUseCase,
        )
    }

    @Test
    fun getBooksWithThumbnailUseCaseTest_GetBooksWithThumbnail_GetBooksWithThumbnailSuccess() {
        runTest {
            val books = getBooksWithThumbnailsUseCase("fake")
            Assert.assertEquals(
                books[0],
                BooksUiModel(id = "id1", title = "title1", url = "thumbnail")
            )
        }
    }
}

File: SwiftUI-Bookshelf-App/Bookshelf appTests/GetBooksWithThumbnailsUseCaseTest.swift
//
//  GetBooksWithThumbnailsUseCaseTest.swift
//  Bookshelf appTests
//
//  Created by Jaehwa Noh on 1/29/24.
//

import XCTest
@testable import Bookshelf_app
import Factory

final class GetBooksWithThumbnailsUseCaseTest: XCTestCase {

    private var getBooksWithThumbnailsUseCase: GetBooksWithThumbnailUseCase?
    
    override func setUpWithError() throws {
        // Put setup code here. This method is called before the invocation of each test method in the class.
        try super.setUpWithError()
        Container.shared.reset()
        Scope.singleton.reset()
        
        Container.shared.booksDataSource.register {
            BooksFakeDataSource()
        }
        
        Container.shared.bookThumbnailDataSource.register {
            ThumbnailFakeDataSource()
        }
        
        getBooksWithThumbnailsUseCase = Container.shared.getBooksWithThumbnailUseCase()
    }

    override func tearDownWithError() throws {
        // Put teardown code here. This method is called after the invocation of each test method in the class.
        try super.tearDownWithError()
        getBooksWithThumbnailsUseCase = nil
    }
    
    func test_GetBooksWithThumbnailUseCaseTest_GetBooksWithThumbnail_GetBooksWithThumbnailSuccess() async throws {
        let books = try await getBooksWithThumbnailsUseCase!(searchTerms: "fake")
        XCTAssertEqual(books[0], BookshelfUiModel(id: "id1", title: "title1", url: "thumbnail"))
    }
}
